consensus/istanbul/backend/announce.go

```diff
@@ -419,3 +516,125 @@ func (sb *Backend) regossipIstAnnounce(msg *istanbul.Message, payload []byte, an
+
+	sb.cachedAnnounceMsgsMu.RLock()
+	defer sb.cachedAnnounceMsgsMu.RUnlock()
+
+	announceVersions := make([]*announceVersion, 0, len(sb.cachedAnnounceMsgs))
+
+	for valAddress, cachedAnnounceEntry := range sb.cachedAnnounceMsgs {
+		announceVersions = append(announceVersions, &announceVersion{ValAddress: valAddress, AnnounceMsgTimestamp: cachedAnnounceEntry.MsgTimestamp})
+	}
+
+	announceVersionsBytes, err := rlp.EncodeToBytes(announceVersions)
+	if err != nil {
+		logger.Error("Error encoding announce versions array", "AnnounceVersions", announceVersions, "err", err)
+		return err
+	}
+
+	logger.Trace("Sending an AnnounceVersions message", "announceVersions", announceVersions, "peer", peer)
+	go peer.Send(istanbulAnnounceVersionsMsg, announceVersionsBytes)
+
+	return nil
+}
+
+// This function will handle a received AnnounceVersions message.
+// Specifically, this node will compare the received version set with it's own,
+// and request announce messages that have a higher version that it's own.
+func (sb *Backend) handleAnnounceVersionsMsg(peer consensus.Peer, payload []byte) error {
+	logger := sb.logger.New("func", "handleAnnounceVersionsMsg", "peer", peer)
+	var announceVersions []announceVersion
+
+	if err := rlp.DecodeBytes(payload, &announceVersions); err != nil {
+		logger.Error("Error in decoding announce versions array", "err", err)
+	}
+
+	logger.Trace("Handling an AnnounceVersions message", "announceVersions", fmt.Sprintf("%v", announceVersions))
+
+	// If the announce's valAddress is not within the registered validator set, then ignore it
+	regAndActiveVals, err := sb.retrieveActiveAndRegisteredValidators()
+	if err != nil {
+		return err
+	}
+
+	announcesToRequest := make([]common.Address, 0)
+	for _, announceVersion := range announceVersions {
+		// Ignore this announceVersion entry if it's val address is not in the current registered or elected valset.
+		if !regAndActiveVals[announceVersion.ValAddress] {
+			logger.Trace("Ignoring announceVersion since it's not in active/elected valset", "valAddress", announceVersion.ValAddress)
+			continue
+		}
+
+		if cachedEntry, ok := sb.cachedAnnounceMsgs[announceVersion.ValAddress]; !ok || cachedEntry.MsgTimestamp < announceVersion.AnnounceMsgTimestamp {
+			announcesToRequest = append(announcesToRequest, announceVersion.ValAddress)
+		}
+	}
+
+	if len(announcesToRequest) > 0 {
+		announcesToRequestBytes, err := rlp.EncodeToBytes(announcesToRequest)
+		if err != nil {
+			logger.Error("Error encoding announce to request array", "announcesToRequest", common.ConvertToStringSlice(announcesToRequest), "err", err)
+			return err
+		}
+
+		logger.Trace("Going to send a GetAnnounces", "announcesToRequest", common.ConvertToStringSlice(announcesToRequest), "peer", peer)
+
+		go peer.Send(istanbulGetAnnouncesMsg, announcesToRequestBytes)
+	}
+
+	return nil
+}
+
+func (sb *Backend) checkPeersAnnounceVersions() {
+	peers := sb.broadcaster.FindPeers(nil, p2p.AnyPurpose)
```

Comment:
This will not include light clients, correct?

Reply:
That's correct.  Only the full nodes/validators.